scripts/rvps-extraction/GetRvps.sh

```diff
@@ -3,60 +3,107 @@
 
 function install_packages() {
     echo "***Installing necessary packages for RVPS values extraction ***"
-    dnf install -y python3 python3-cryptography kmod 
+    dnf install -y python3 python3-cryptography kmod
     echo "***Installation Finished ***"
 }
 
 # Function to mount the image and extract se.img
 function mount_and_extract_image() {
     local img_path=$1
-    
+
     # Cleanup any previous files and directories
     rm -rf se.img /mnt/myvm
     mkdir /mnt/myvm
-    
+
     # Load nbd module and mount the image
     modprobe nbd
     if [ $? -ne 0 ]; then
         echo "Error: Failed to load nbd module."
         exit 1
     fi
 
-    qemu-nbd -c /dev/nbd3 $img_path
+
+    #Check the available nbd and use it
+    nbd_available=`lsblk | grep nbd | grep 0B | awk '{print $1}' | head -1`
+    echo $nbd_available "looks available. Starting Mounting the image with it.."
+    qemu-nbd -c /dev/$nbd_available $img_path
+
     if [ $? -ne 0 ]; then
         echo "Error: Failed to connect to nbd device."
         exit 1
     fi
 
 
-    mount /dev/nbd3p1 /mnt/myvm
+
+    mount /dev/"$nbd_available"p1 /mnt/myvm
     if [ $? -ne 0 ]; then
          echo "Error: Failed to mount the image. Retrying..."
          sleep 2
-         mount /dev/nbd3p1 /mnt/myvm
+         mount /dev/"$nbd_available"p1 /mnt/myvm
          if [ $? -ne 0 ]; then
-     		echo "Retrial for mounting failed. Please rerun the script"
-     		exit 1
-	 else
-		echo "Mounting on second attempt passed"
-         fi
-     fi
+	      echo "Retrial for mounting failed. Creating "$nbd_available"p1 manually.."
+             #Parsing lsblk output to get the major and minor numbers assigned to nbd.
+              devmajor=$(lsblk | grep -w "$nbd_available" | awk '{print $2}' | awk -F ":" '{print $1}' | xargs)
+              devminor=$(lsblk | grep -w "$nbd_available" | awk '{print $2}' | awk -F ":" '{print $2}' | xargs)
+
+              # Ensure devminor is a number before performing arithmetic
+              # The device minor version form nbdxp1 should be equal to minor_version of nbdx + 1
+              if [[ "$devminor" =~ ^[0-9]+$ ]]; then
+                    newdevminor=$((devminor + 1))
+              else
+	            echo $devminor
+                    echo "Error: nbd minor value is not a valid integer. Disconnecting and exiting.."
+	            qemu-nbd -d /dev/"$nbd_available"
+                    exit 1
+              fi
+
+	      sleep 2
+	      #manual creation of nbdxp1 using mknod
+	      mknod /dev/"$nbd_available"p1 b $devmajor $newdevminor
```

Comment:
From the manual page of `qemu-nbd` :

```
Expose  the  guest-visible  contents of a qcow2 file via a block device
/dev/nbd0 (and possibly creating /dev/nbd0p1 and friends for partitions
found within), then disconnect the device when done.
[...]
qemu-nbd -c /dev/nbd0 -f qcow2 file.qcow2
```

Why are you trying to create the device node manually ? How do you expect this would help ?

Reply:
hi @gkurz , we faced few cases where nbd0 is engaged and thus get error while using it. e.g :- 
Refer to below case - 
https://docs.redhat.com/en/documentation/openshift_sandboxed_containers/1.8/html/user_guide/deploying-cc_ibm-cc#ibm-cc-obtaining-attest-fields-certs-keys_ibm-cc:~:text=%24%20lsblk-,Example%20output,-nbd0%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2043%3A0


Then in this case, we can see 'nbd0p1' is engaged by 'luks'  - 
nbd0                                           43:0    0  100G  0 disk
├─nbd0p1                                       43:1    0  255M  0 part
├─nbd0p2                                       43:2    0    6G  0 part
│ └─luks-e23e15fa-9c2a-45a5-9275-aae9d8e709c3 253:2    0    6G  0 crypt
└─nbd0p3                                       43:3    0 12.4G  0 part
nbd1                                           43:32   0   20G  0 disk
├─nbd1p1                                       43:33   0  255M  0 part
├─nbd1p2                                       43:34   0    6G  0 part
│ └─luks-5a540f7c-c0cb-419b-95e0-487670d91525 253:3    0    6G  0 crypt
└─nbd1p3                                       43:35   0 86.9G  0 part
nbd2                                           43:64   0    0B  0 disk
nbd3                                           43:96   0    0B  0 disk
nbd4                                           43:128  0    0B  0 disk
nbd5                                           43:160  0    0B  0 disk
nbd6                                           43:192  0    0B  0 disk
nbd7                                           43:224  0    0B  0 disk
nbd8                                           43:256  0    0B  0 disk
nbd9                                           43:288  0    0B  0 disk
nbd10                                          43:320  0    0B  0 disk

So with the new logic, next available 'nbd' is 'nbd2' and it will be used.

If 'nbd0' will be available , then surely it will be used.

Please let me know if it answers your query.

Reply:
> Hi @gkurz,
> 
> We encountered a few cases where nbd0 is engaged, resulting in errors when attempting to use it. For example:
> 
> [...]
> 
> With the new logic, the next available nbd device is nbd2, and it will be used. If nbd0 becomes available, it will certainly be used.
> 

As already stated in https://github.com/openshift/sandboxed-containers-operator/pull/486#pullrequestreview-2498995978, I'm fine with this part.

> Additionally, we faced some cases where nbd0p1 could not be used due to a partition error. In such scenarios, the script attempts to handle the issue by manually creating the partition. I’ve implemented the following logic:
> 
> 1. Attempt to mount: mount /dev/"$nbd_available"p1 /mnt/myvm (This is the first attempt).
> 2. Retry: If step 1 fails, the script retries once.
> 3. Create manually: If the retrial also fails (a rare occurrence), the script creates the device using mknod /dev/"$nbd_available"p1 b $devmajor $newdevminor.
> 

This doesn't make sense. The `mknod` command simply creates a block device special file, which is basically an interface between userspace and the nbd kernel driver. It will not create a partition in any way.

> In most cases, step 1 works. In a few cases, we encountered step 2, so I added a retry mechanism. Finally, in one isolated scenario on an LPAR, we faced step 3, so I included that logic as well.
> 

The nbd server runs in userspace. The on-lining of the nbd block device is asynchronous. It is certainly possible that it takes some time to complete inside an overloaded LPAR.

> Essentially, I have accounted for all these scenarios to ensure the script helps users retrieve the required values seamlessly.
> 

Step 3 is a no go. The only thing to do is to try again until the `/dev/nbdXp1`  eventually shows up or bail out after a certain amount of time.

> Please let me know if this addresses your query. I will share additional details if needed.

Reply:
Thanks @gkurz . I have removed that part.

Result looks good to me. I am sharing below - 

Script Result from PODVM volume image :- 
```
1) Generate the RVPS From Local Image from User pc
2) Generate RVPS from Volume
3) Quit
Please enter your choice: 2
Enter the Libvirt Pool Name
fa-pp
Enter the Libvirt URI Name
qemu+ssh://root@192.168.122.1/system?no_verify=1
Enter the Libvirt Volume Name
fa-pp-vol
Downloading from PODVM Volume...

nbd0 looks available. Starting Mounting the image with it..
/dev/nbd0 disconnected
SE header found at offset 0x014000
SE header written to '/root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/hdr.bin' (640 bytes)
se.tag:  42f3fe61e8a7e859cab3bb033fd11c61
se.image_phkh:  92d0aff6eb86719b6b1ea0cb98d2c99ff2ec693df3efff2158f54112f6961508
provenance = ewogICAgInNlLmF0dGVzdGF0aW9uX3Boa2giOiBbCiAgICAgICAgIjkyZDBhZmY2ZWI4NjcxOWI2YjFlYTBjYjk4ZDJjOTlmZjJlYzY5M2RmM2VmZmYyMTU4ZjU0MTEyZjY5NjE1MDgiCiAgICBdLAogICAgInNlLnRhZyI6IFsKICAgICAgICAiNDJmM2ZlNjFlOGE3ZTg1OWNhYjNiYjAzM2ZkMTFjNjEiCiAgICBdLAogICAgInNlLmltYWdlX3Boa2giOiBbCiAgICAgICAgIjkyZDBhZmY2ZWI4NjcxOWI2YjFlYTBjYjk4ZDJjOTlmZjJlYzY5M2RmM2VmZmYyMTU4ZjU0MTEyZjY5NjE1MDgiCiAgICBdLAogICAgInNlLnVzZXJfZGF0YSI6IFsKICAgICAgICAiMDAiCiAgICBdLAogICAgInNlLnZlcnNpb24iOiBbCiAgICAgICAgIjI1NiIKICAgIF0KfQo=
-rw-r--r--. 1 root root 640 Dec 13 12:13 /root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/hdr.bin
-rw-r--r--. 1 root root 446 Dec 13 12:13 /root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/ibmse-policy.rego
-rw-r--r--. 1 root root 561 Dec 13 12:13 /root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/se-message
```



Script Result from PODVM local image :-

```
1) Generate the RVPS From Local Image from User pc
2) Generate RVPS from Volume
3) Quit
Please enter your choice: 1
Enter the Qcow2 image with Full path
/root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/PODVM-VOL-IMAGE/podvm_test.qcow2
nbd0 looks available. Starting Mounting the image with it..
mount: /mnt/myvm: special device /dev/nbd0p1 does not exist.
Error: Failed to mount the image. Retrying...
Mounting on second attempt passed
/dev/nbd0 disconnected
SE header found at offset 0x014000
SE header written to '/root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/hdr.bin' (640 bytes)
se.tag:  42f3fe61e8a7e859cab3bb033fd11c61
se.image_phkh:  92d0aff6eb86719b6b1ea0cb98d2c99ff2ec693df3efff2158f54112f6961508
provenance = ewogICAgInNlLmF0dGVzdGF0aW9uX3Boa2giOiBbCiAgICAgICAgIjkyZDBhZmY2ZWI4NjcxOWI2YjFlYTBjYjk4ZDJjOTlmZjJlYzY5M2RmM2VmZmYyMTU4ZjU0MTEyZjY5NjE1MDgiCiAgICBdLAogICAgInNlLnRhZyI6IFsKICAgICAgICAiNDJmM2ZlNjFlOGE3ZTg1OWNhYjNiYjAzM2ZkMTFjNjEiCiAgICBdLAogICAgInNlLmltYWdlX3Boa2giOiBbCiAgICAgICAgIjkyZDBhZmY2ZWI4NjcxOWI2YjFlYTBjYjk4ZDJjOTlmZjJlYzY5M2RmM2VmZmYyMTU4ZjU0MTEyZjY5NjE1MDgiCiAgICBdLAogICAgInNlLnVzZXJfZGF0YSI6IFsKICAgICAgICAiMDAiCiAgICBdLAogICAgInNlLnZlcnNpb24iOiBbCiAgICAgICAgIjI1NiIKICAgIF0KfQo=
-rw-r--r--. 1 root root 640 Dec 13 12:14 /root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/hdr.bin
-rw-r--r--. 1 root root 446 Dec 13 12:14 /root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/ibmse-policy.rego
-rw-r--r--. 1 root root 561 Dec 13 12:14 /root/Rvps-Extraction/sandboxed-containers-operator/scripts/rvps-extraction/output-files/se-message
```

Reply:
Hi @gkurz , done. Please can you check.

Reply:
@gauravkuredhat please add commit log and Signed-off-by.

Reply:
Done @gkurz